netplan/libnetplan.py

```diff
@@ -18,19 +18,167 @@
 
 import tempfile
 import logging
+from collections import defaultdict
 import ctypes
 import ctypes.util
 from ctypes import c_char_p, c_void_p, c_int, c_uint, c_size_t, c_ssize_t
 from typing import List, Union, IO
+from enum import IntEnum
 from io import StringIO
 import os
+import re
 
 
-class LibNetplanException(Exception):
+# Errors and error domains
+
+# NOTE: if new errors or domains are added,
+# include/types.h must be updated with the new entries
+class NETPLAN_ERROR_DOMAINS(IntEnum):
+    NETPLAN_PARSER_ERROR = 1
+    NETPLAN_VALIDATION_ERROR = 2
+    NETPLAN_FILE_ERROR = 3
+    NETPLAN_BACKEND_ERROR = 4
+    NETPLAN_EMITTER_ERROR = 5
+    NETPLAN_FORMAT_ERROR = 6
+
+
+class NETPLAN_PARSER_ERRORS(IntEnum):
+    NETPLAN_ERROR_INVALID_YAML = 0
+    NETPLAN_ERROR_INVALID_CONFIG = 1
+
+
+class NETPLAN_VALIDATION_ERRORS(IntEnum):
+    NETPLAN_ERROR_CONFIG_GENERIC = 0
+    NETPLAN_ERROR_CONFIG_VALIDATION = 1
+
+
+class NETPLAN_BACKEND_ERRORS(IntEnum):
+    NETPLAN_ERROR_UNSUPPORTED = 0
+    NETPLAN_ERROR_VALIDATION = 1
+
+
+class NETPLAN_EMITTER_ERRORS(IntEnum):
+    NETPLAN_ERROR_YAML_EMITTER = 0
+
+
+class NETPLAN_FORMAT_ERRORS(IntEnum):
+    NETPLAN_ERROR_FORMAT_INVALID_YAML = 0
+
+
+class NetplanException(Exception):
+    def __init__(self, message=None, domain=None, error=None):
+        self.domain = domain
+        self.error = error
+        self.message = message
+
+    def __str__(self):
+        return self.message
+
+
+class NetplanFileException(NetplanException):
     pass
 
 
-class _GError(ctypes.Structure):
+class NetplanValidationException(NetplanException):
+    '''
+    Netplan Validation errors are expected to contain the YAML file name
```

Comment:
Can this docstr explain in what cases we expect this class to be used (and a docstr on `NetplanParserException` to better differentiate what symptom each exception indicates).